src/unittest/test_mapgen.cpp

```diff
@@ -0,0 +1,124 @@
+// Luanti
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2024 Minetest core developers & community
+
+#include "test.h"
+
+#include "emerge.h"
+#include "mapgen/mapgen.h"
+#include "mapgen/mg_biome.h"
+#include "mock_server.h"
+
+class TestMapgen : public TestBase
+{
+public:
+	TestMapgen() { TestManager::registerTestModule(this); }
+	const char *getName() { return "TestMapgen"; }
+
+	void runTests(IGameDef *gamedef);
+
+	void testMapgen(IGameDef *gamedef);
+};
+
+static TestMapgen g_test_instance;
+
+namespace {
+	class MockBiomeManager : public BiomeManager {
+	public:
+		MockBiomeManager(Server *server) : BiomeManager(server) {}
+
+		void setNodeDefManager(const NodeDefManager *ndef)
+		{
+			m_ndef = ndef;
+		}
+	};
+}
+
+void TestMapgen::runTests(IGameDef *gamedef)
+{
+	TEST(testMapgen, gamedef);
+}
+
+void TestMapgen::testMapgen(IGameDef *gamedef)
+{
+	MockServer server(getTestTempDirectory());
+	MockBiomeManager bmgr(&server);
+	bmgr.setNodeDefManager(gamedef->getNodeDefManager());
+
+	enum FakeNodes : content_t {
+		FN_DIRT_WITH_GRASS = 10,
+		FN_DIRT,
+		FN_STONE,
+	};
+
+	{
+		// Add some biomes (equivalent to l_register_biome)
+		// Taken from minetest_game/mods/default/mapgen.lua
+		size_t bmgr_count = bmgr.getNumObjects(); // this is 1 ?
+
+		Biome *b = BiomeManager::create(BIOMETYPE_NORMAL);
+		b->name = "deciduous_forest";
+		b->c_top = FN_DIRT_WITH_GRASS;
+		b->depth_top = 1;
+		b->c_filler = FN_DIRT;
+		b->depth_filler = 3;
+		b->c_stone = FN_STONE;
+		b->min_pos.Y = 1;
+		b->heat_point = 60.0f;
+		b->humidity_point = 68.0f;
+		UASSERT(bmgr.add(b) != OBJDEF_INVALID_HANDLE);
+
+		b = BiomeManager::create(BIOMETYPE_NORMAL);
+		b->name = "deciduous_forest_shore";
+		b->c_top = FN_DIRT;
+		b->depth_top = 1;
+		b->c_filler = FN_DIRT;
+		b->depth_filler = 3;
+		b->c_stone = FN_STONE;
+		b->max_pos.Y = 0;
+		b->heat_point = 60.0f;
+		b->humidity_point = 68.0f;
+		UASSERT(bmgr.add(b) != OBJDEF_INVALID_HANDLE);
+		UASSERT(bmgr.getNumObjects() - bmgr_count == 2);
+	}
+
+
+	std::unique_ptr<BiomeParams> params(BiomeManager::createBiomeParams(BIOMEGEN_ORIGINAL));
+
+	constexpr v3s16 CSIZE(16, 16, 16);
```

Comment:
note: this is usually 5

Reply:
It's unfortunate naming. It seems to be in nodes.
https://github.com/minetest/minetest/blob/ae96a8d4fa9ade70672fe8d2152013c547edfb47/src/emerge.cpp#L193-L194

Reply:
ah, then 80